src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,212 @@
+            return oldNode.value;
+        }
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        if (size == 0) {
+            return false;
+        }
+        int index = -1;
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == object || (currentNode.value != null
+                    && currentNode.value.equals(object))) {
+                index = i;
+                break;
+            }
+            currentNode = currentNode.next;
+        }
+        if (index == -1) {
+            return false;
+        }
+        if (index == 0) {
+            removeFirst();
+            return true;
+        } else if (index == size - 1) {
+            removeLast();
+            return true;
+        } else {
+            unlink(currentNode);
+            size--;
+            return true;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    @Override
+    public String toString() {
+        return "MyLinkedList{"
+                + "size=" + size
+                + ", head=" + head.value
+                + ", tail=" + tail.value
+                + '}';
+    }
+
+    private class Node<T> {
+        private Node<T> prev;
+        private T value;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
+    private Node<T> findElement(int index) {
+        Node<T> currentNode;
+        if (index < size / 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    private void addFirst(T value) {
+        Node<T> oldNode = head;
+        head = new Node<T>(null, value, oldNode);
+        if (tail == null) {
+            tail = head;
+        } else {
+            oldNode.prev = head;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        Node<T> oldNode = tail;
+        tail = new Node<T>(oldNode, value, null);
+        if (head == null) {
+            head = tail;
+        } else {
+            oldNode.next = tail;
+        }
+        size++;
+    }
+
+    private void unlink(Node<T> node) {
+        node.next.prev = node.prev;
+        node.prev.next = node.next;
+        node.prev = null;
+        node.next = null;
+    }
+
+    private void checkIndex(int index, int limit) {
+        if (index < 0 || index > limit) {
+            throw new IndexOutOfBoundsException("Index is outside");
```

Comment:
```suggestion
            throw new IndexOutOfBoundsException("Index " + index + " is out or range for limit " + limit);
```

lets make exception messages more informative